src/neo/Network/P2P/Payloads/ExtensiblePayload.cs

```diff
@@ -0,0 +1,94 @@
+using Neo.IO;
+using Neo.Ledger;
+using Neo.Persistence;
+using Neo.SmartContract;
+using System;
+using System.IO;
+
+namespace Neo.Network.P2P.Payloads
+{
+    public class ExtensiblePayload : IInventory
+    {
+        public string Receiver;
+        public uint ValidBlockStart;
+        public uint ValidBlockEnd;
+        public byte[] Data;
+        public Witness Witness;
+
+        private UInt256 _hash = null;
+        public UInt256 Hash
+        {
+            get
+            {
+                if (_hash == null)
+                {
+                    _hash = this.CalculateHash();
+                }
+                return _hash;
+            }
+        }
+
+        InventoryType IInventory.InventoryType => InventoryType.Extensible;
+
+        public int Size =>
+            Receiver.GetVarSize() + //Receiver
+            sizeof(uint) +          //ValidBlockStart
+            sizeof(uint) +          //ValidBlockEnd
+            Data.GetVarSize() +     //Data
+            1 + Witness.Size;       //Witness
+
+        Witness[] IVerifiable.Witnesses
+        {
+            get
+            {
+                return new[] { Witness };
+            }
+            set
+            {
+                if (value.Length != 1) throw new ArgumentException();
+                Witness = value[0];
+            }
+        }
+
+        void ISerializable.Deserialize(BinaryReader reader)
+        {
+            ((IVerifiable)this).DeserializeUnsigned(reader);
+            if (reader.ReadByte() != 1) throw new FormatException();
+            Witness = reader.ReadSerializable<Witness>();
+        }
+
+        void IVerifiable.DeserializeUnsigned(BinaryReader reader)
+        {
+            Receiver = reader.ReadVarString(32);
+            ValidBlockStart = reader.ReadUInt32();
+            ValidBlockEnd = reader.ReadUInt32();
+            Data = reader.ReadVarBytes(ushort.MaxValue);
+        }
+
+        UInt160[] IVerifiable.GetScriptHashesForVerifying(StoreView snapshot)
+        {
+            return new[] { Witness.ScriptHash }; // This address should be checked by consumer
+        }
```

Comment:
Add `Sender` field? Otherwise `ContractParametersContext` doesn't work.

Reply:
`public UInt160 Sender => Witness.ScriptHash`?

Reply:
Why we need `Sender`?

Reply:
1. In `GetScriptHashesForVerifying()`, you access `Witnesses`.
2. To set the `Witnesses`, you have to use `ContractParametersContext` to sign the payload.
3. The `ContractParametersContext` use `GetScriptHashesForVerifying()` to get the hashes.
4. Then you enter a loop.

Reply:
Done